quesma/quesma/router.go

```
@@ -84,75 +84,47 @@ func configureRouter(cfg config.QuesmaConfiguration, sr schema.Registry, lm *cli
 
-					return true
-				})
+		normalizedPattern := elasticsearch.NormalizePattern(req.Params["index"])
 
-			return resolveIndexResult(sources), nil
-		} else {
-			if config.MatchName(elasticsearch.NormalizePattern(pattern), pattern) {
-				definitions, err := lm.GetTableDefinitions()
-				if err != nil {
-					return nil, err
-				}
-
-				if definitions.Has(pattern) {
-					return resolveIndexResult(elasticsearch.Sources{
-						Indices: []elasticsearch.Index{},
-						Aliases: []elasticsearch.Alias{},
-						DataStreams: []elasticsearch.DataStream{
-							{
-								Name:           pattern,
-								BackingIndices: []string{pattern},
-								TimestampField: `@timestamp`,
-							},
-						},
-					}), nil
-				} else {
-					return &mux.Result{StatusCode: 404}, nil
-				}
-			} else {
-				sources, found, err := elasticsearch.NewIndexResolver(cfg.Elasticsearch.Url.String()).Resolve(pattern)
-				if err != nil {
-					return nil, err
-				}
-				if !found {
+		// Optimization: if it's not a pattern, let's try avoiding querying Elasticsearch - let's first try
+		// finding that index in schema.Registry:
+		if !elasticsearch.IsIndexPattern(normalizedPattern) {
+			if foundSchema, found := sr.FindSchema(schema.TableName(normalizedPattern)); found {
+				if !foundSchema.ExistsInDataSource {
+					// index configured by the user, but not present in the data source
 					return &mux.Result{StatusCode: 404}, nil
 				}
 
-				return resolveIndexResult(sources), nil
+				return resolveIndexResult(elasticsearch.Sources{
+					Indices: []elasticsearch.Index{},
+					Aliases: []elasticsearch.Alias{},
+					DataStreams: []elasticsearch.DataStream{
+						{
+							Name:           normalizedPattern,
+							BackingIndices: []string{normalizedPattern},
+							TimestampField: `@timestamp`,
+						},
+					},
+				}), nil
 			}
+
+			// ...index not found in schema.Registry (meaning the user did not configure it), so let's fall back to
```

I think "fall back" term might be slightly misleading - we don't really "fall back" because the first one did not work - we simply router to Elastic because no one explicitly stated that this particular index name should be resolved against Clickhouse

---

Changed the wording to `...index not found in schema.Registry (meaning the user did not configure it), but it could exist in Elastic`